builder/builder.go

```diff
@@ -522,137 +522,135 @@ func (b *Builder) BuildChroots(template *x509.Certificate, privkey *rsa.PrivateK
 	return nil
 }
 
-// Set the published versions to what was just built
-func (b *Builder) setVersion(publish bool) {
-	if publish == false {
-		return
-	}
-
-	// Create the www/version/format# dir if it doesn't exist
-	formatdir := b.Statedir + "/www/version/format" + b.Format
-	if _, err := os.Stat(formatdir); os.IsNotExist(err) {
-		if err = os.MkdirAll(formatdir, 0777); err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
-		}
+// BuildUpdate will produce an update consumable by the swupd client
+func (b *Builder) BuildUpdate(prefixflag string, minVersion int, format string, skipSigning bool, publish bool, keepChroots bool) error {
+	if format != "" {
+		b.Format = format
 	}
 
-	fmt.Println("Setting latest version to " + b.Mixver)
-	err := ioutil.WriteFile(formatdir+"/latest", []byte(b.Mixver), 0644)
+	// Ensure the format dir exists.
+	formatDir := filepath.Join(b.Statedir, "www", "version", "format"+b.Format)
+	err := os.MkdirAll(formatDir, 0777)
 	if err != nil {
-		helpers.PrintError(err)
-		os.Exit(1)
+		return errors.Wrapf(err, "couldn't create the format directory")
 	}
 
-	err = ioutil.WriteFile(b.Statedir+"/image/LAST_VER", []byte(b.Mixver), 0644)
+	//
+	// Create update metadata for the mix.
+	//
+	updatecmd := exec.Command(prefixflag+"swupd_create_update", "-S", b.Statedir, "--minversion", strconv.Itoa(minVersion), "-F", b.Format, "--osversion", b.Mixver)
+	updatecmd.Stdout = os.Stdout
+	updatecmd.Stderr = os.Stderr
+	err = updatecmd.Run()
 	if err != nil {
-		helpers.PrintError(err)
-		os.Exit(1)
+		return errors.Wrapf(err, "failed to create update metadata")
 	}
 
-	if b.Upstreamurl != "" {
-		fmt.Println("Saving the upstream version URL " + b.Upstreamurl)
-		upstreamurl := b.Statedir + "/www/" + b.Mixver + "/upstreamurl"
-		err = ioutil.WriteFile(upstreamurl, []byte(b.Upstreamurl), 0644)
+	// Clean up the bundle chroots as only the full chroot is needed from this point on.
+	if !keepChroots {
+		var files []os.FileInfo
+		files, err = ioutil.ReadDir(b.Bundledir)
 		if err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
+			fmt.Fprintf(os.Stderr, "Ignored error when cleaning bundle chroots: %s", err)
+		}
+		basedir := filepath.Join(b.Statedir, "image", b.Mixver)
+		for _, f := range files {
+			if f.Name() == "full" {
+				continue
+			}
+			err = os.RemoveAll(filepath.Join(basedir, f.Name()))
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Ignored error when cleaning bundle chroots: %s", err)
```

Comment:
@phmccarty @matthewrsj Is this something worth printing? Do we care whether or not those files really got removed?

Reply:
I think this would work better as a `TODO` comment in the code if we should be caring about the errors.

Reply:
Mixer says it will clean those (unless you pass a flag), but it couldn't. Since the process can proceed, we just warn the user about the problem.

Reply:
We could print at the end instead of in the loop as well, to prevent spamming things. This would be interesting if it really failed though, @cmarcelo we can keep this and see how it goes if people actually run into it.